ARCs/arc-draft_alternative_0014.md

```diff
@@ -0,0 +1,285 @@
+2. Bob generates a message `msg` for *PKa* and request Alice to sign it;
+3. Alice signs `msg` using her wallet, and afterwards sends back the tuple `<PKa, Sig(msg)>` to Bob;
+4. Bob verifies the digital signature `Sig(msg)` with the *PKa* and `msg`;
+5. If the signature is valid, then Bob generates a new `session-id` for Alice; otherwise returns NULL (authentication rejected);
+
+An ARC-0014 compliant message is called *Authentication Message*. A User **MUST** sign an *Authentication Message* with the *SKa* stored into a wallet.
+
+### Authentication Message
+
+An Authentication Message is a sequence of bytes representing a message. The Verifier asks Users to sign an *Authentication Message* with their secret keys. Such a message **MUST** include the following information:
+
+- `domain name` of the service;
+- `description` of the service;
+- `Algorand address` (PKa) to be authenticated;
+- `nonce` representing a unique/random value generated by the verifier.
+
+The JSON structure for such an object is:
+
+```typescript
+interface AuthMessage {
+ /** The domain name of the service */
+ service: string;
+ /** Optional, description of the service */
+ desc?: string;
+ /** Algorand account to authenticate with*/
+ authAcc: string;
+ /** Challenge generated server-side */
+ nonce: string;
+}
+```
+
+For example:
+
+```json
+{
+ "service": "www.servicedomain.com",
+ "desc": "Domain offers important services to users",
+ "authAcc": "KTGP47G64KCXWJS64W7SGJNKTHE37TYDCI64USXI3XOYE6ZSH4LCI7NIDA",
+ "nonce": "1234abcde!%",
+}
+```
+
+The `nonce` field **MUST** be unique for each authentication and **MUST NOT** be used more than once to avoid replay attacks.
+
+Users sign *Authentication Message*s using the secret keys stored into a wallet.
+
+***Problem***: Most of the Algorand wallets can only sign Algorand `Transaction` objects, and do not offer the possibility to sign arbitrary bytes. To overcome such a limitation, the ARC-0014 distinguishes two authentication objects, namely the **Simple Authentication Message**, and the **Transaction Authentication Message**. The former **SHOULD** be used by any signing mechanisms able to sign random bytes, whereas the latter can be used with almost any Algorand wallet.
+
+#### Simple Authentication Message
+
+The *Simple Authentication Message* is a sequence of bytes representing an ARC-0014 *Authentication Message*. It **SHOULD** be used in contexts that support signatures of random bytes. A *Simple Authentication Message* is an *Authentication Message* prepended with the prefix `arc14` for domain separation. It **MUST** be represented as the hash SHA-512/256 of such prefix, together with a msgpack encoded *Authentication Message*. For example, given an *Authentication Message* object `aut_message`, its *Simple Authentication Message* representation would be something like: `SHA512_256(“arc14”+msgpacked_auth_message)`.
+
+#### Transaction Authentication Message
```

Comment:
What's the rationale for creating a transaction-based implementation alongside the message-based implementation? Short term I see the value of being "supported" by wallets right now since most don't support message signing. However, I think the main reason we want this standard is so wallets have a standard to follow when implementing an authentication UI, thus giving the end-user a better experience. If proper support of this ARC involves software updates regardless of implementation, then requiring message signing is not much more costly from a technical perspective.

Also, from a dev UX perspective, if a wallet has a auth endpoint, it'd be confusing if some wallets expect a transaction and some wallets expect a message. It'd be better if all wallets had the same standardized input/output.

Reply:
I agree. I can understand the value of being backwards-compatible, but I think, in the long run, supporting a single type of message can be better for the ecosystem.

Reply:
I deffo agree that in the long term, wallets will tend to support message signing. However, the problem I see is that this feature is still not offered by most of them. This is a huge limitation for dApp devs that want to implement web3 authentication mechanisms. 

I am still convinced that ARC13 must be inclusive, offering a standard approach for any kind of wallet. It should not force users to adopt only wallets offering message signing features. In addition, backward compatibility is important and cannot be ignored.

IMHO I think the ARC must offer a standard for both Transaction/Message signing approaches and let users decide which wallet and approach to use. It is then a duty of wallets to provide guidelines on which arc14 authentication they support.

Reply:
I think we should be writing ARCs for the wallets to follow, not the other way around. If a wallet doesn't have a proper auth endpoint/UI, having a auth-specific transaction isn't very helpful. End-users that are not technically savvy enough to understand why they are signing a transaction will not bother reading the note field and comparing it to the ARC, so you're effectively back to square. If a wallet does implement an auth UI, implementing functionality for message-signing is trivial since it is a native feature supported by SDKs.

Reply:
Ok what is your suggestion then? Should we go just with the _simple authentication message_? This will cut off most of today's wallets (maybe all of them if I am not wrong), making ARC14 a good standard, but impractical for now.

I'd also like to have comments from @yigitguler, @barnjamin, and @cusma on that.

Reply:
This seems like a security consideration for signing arbitrary bytes, tag @fabrice102

Reply:
web3 projects are already working with "workaround" authentication solutions. I think there is a great value of having a real "standard" where dApps can trust that their users will be authenticated with ease, with wallets that support ARC-14. I think wallet developers will be willing to implement ARC-14 as quickly as possible since it will make their users safer.

Reply:
We decided to keep both Simple Authentication Message or Transaction Authentication Message in ARC14. Wallets will handle the UX showing wich kind of message they support.
@SudoWeezy I am solving this thread for now.